### DISCUSSION TITLE
Grayed out sets when lis_source_did is empty

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork2#2582] [General]
If a course's grade passback mode is `homework`, but a set's `lis_source_did` has not yet been established, the set is grayed out. I think I understand the reason for this is that the student must have entered webwork some other way (password authentication, some other LTI link into the course, ...) and it leads to the student telling the instructor something is wrong, and eventually it is understood that someone needs to follow the LTI link from the LMS to get that `lis_source_did` recorded.

I wanted to ask if we really need to gray out the set under these conditions. Correct me if I'm wrong, but if the set were not grayed out, the student could work on it, and the only "bad" thing is that grade passback will not occur. Wouldn't that be OK? And later the instructor can recognize grade passback is not happening and do something to establish the value for the set's `lis_source_did`?  Or maybe even if the grade passback mode is `homework` and `lis_source_did` is not yet established, then there would be an alert next to the set on the Assignments page for the instructor to see.

It's been a problem repeatedly that instructors who do not grok LTI well are reporting students unable to access their sets because of this graying out.

Additionally/alternatively, at the time an instructor in the LMS uses the content item selection tool and data is sent to WeBWorK to retrieve the list of sets, is enough known already such that the `lis_source_did` for all sets could be set in the database at that time?

I'm asking these questions to make sure I don't misunderstand something. I may make some changes to the LTI if I understand things correctly. Maybe I'd make the graying out behavior optional if I understand things correctly, but people would still want that behavior.

**Comment by drgrice1** [comment]
One of the problems is that it is not easy for an instructor to "establish the value for the set's `lis_source_did`".  It is different for LTI 1.1 than for LTI 1.3.  For LTI 1.1, the only way that can really be established is by the student entering the set from the LMS.  There may be ways to delve into the LMS's database and find what is needed, but that is not really a viable option.  For LTI 1.3, it still requires the student entering a set from the LMS, but it is more broad as once any student has entered a set, it enables the set for all students that have at least signed in once from the LMS.  However, as with LTI 1.1, if none of the required passback information is available, then it isn't easy to find in any other way than a student entering WeBWorK from the LMS.

As to using content item selection (or deep linking), the answer is no, that information is not available at that time.

**Reply by Alex-Jordan** [reply]
For us with LTI 1.3 with D2L, it is enough for the instructor to follow a set's link from the LMS. My practice has become to create one module where I use the content item selection to create links to all of the sets. I hide this module from students, although I could reveal it I suppose. But I then just open all the links myself once, and that's enough for students.

Elsewhere in the course I can get away with only providing a link to the WW course landing page. Or I can make additional instances of the set-specific links within the relevant modules (but I need to use the already-existing LTI/ELT links, not use the content item selection tool again). It doesn't matter if a student uses the set-specific links since I already initialized things.

But it is this step where the instructor would open all the links that my faculty keep overlooking. It's in my instructions and my tutorial video for them, but enough of them are just not doing it for some reason.

**Comment by drgrice1** [comment]
That may work for D2L (and perhaps for Moodle), but will not work for all LMS's.  It specifically will not work for Canvas, as it does not provide grade pass back information for instructors.

**Comment by drgrice1** [comment]
Also, it won't work for LTI 1.1 at all since each user has set specific information that must be provided.

**Comment by Alex-Jordan** [comment]
So it sounds like it would be fine to make it configurable whether or not the sets under those conditions are grayed out. The default would be as things are now.  But for example with D2L and LTI 1.3, changing this setting would make sense for some instances. Does that all sound right to you?

**Comment by drgrice1** [comment]
Yeah, that seems fine to me.

**Comment by dlglin** [comment]
I've been running into the same thing as @Alex-Jordan this term, so I'd love to have the behaviour be configurable.

**Comment by Alex-Jordan** [comment]
Even when the set is grayed out, we should make the message more clear to a student. Right now, in the mind of a student, what it says is compatible with go to the LMS, use some link to enter WeBWorK, and then go to the set. In other words, it can be understood to mean all that is required is that you enter WeBWorK from the LMS. Not that you need to specifically use an LTI link that goes straight to this assignment.

This just came up again this week so I am motivated to try a patch on my production server. I will open a PR if what I get going is working and looks OK.

**Comment by Alex-Jordan** [comment]
I am working on this. Suppose the configuration is such that LTI 1.3 is in use and students are not blocked from accessing a set just because the lis_source_did is missing.

What should we do to alert the instructor that the lis_source_did is missing? One option is that in Sets Manager, there is some indication next to each such set. Can anyone suggest a better option?

**Comment by Alex-Jordan** [comment]
Note that using Sets Manager, as opposed to the Assignments page or something else, means there's no need for some new permission. (You wouldn't want this cluttering up student screens, so you'd only show it to users at a new permission level.) If it's in Sets Manager, it can just be show all the time assuming LTI 1.3 is in use. (And maybe also assuming the new config option is toggled.)

**Comment by drgrice1** [comment]
If you can fit it onto the Sets Manager page in a nice way, then that seems like a good place to put it.

**Comment by dlglin** [comment]
@Alex-Jordan did this get implemented?  I thought I remembered a PR, but maybe I'm wrong.

**Comment by Alex-Jordan** [comment]
No, I do not recall actually getting to work on this.

**Comment by Alex-Jordan** [comment]
@dlglin I don't think I will have time for this for 2.20. I haven't looked, but I am worried there are many places that will need attention. Not just the page with the grayed out sets, but also other places where access to a set is blocked. If you are interested in working on this, I think the only way it will happen for 2.20 is if you tackle it. We can always carry on as we have been...

**Comment by somiaj** [comment]
A new option to override this was introduced and all sets can be entered now, but messaging is now given to students who need access the assignment though their LTI first.
